### PR TITLE
Add expand operator

### LINKED ISSUE #4029: Expand Operator.

### PR DESCRIPTION
**Pull request by pkuyym**
Resolves #4029

### CODE REVIEW

File: paddle/operators/expand_op.h [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    break;                                                         \
+  }
+#define EXPAND_TEMPLATE_GRAD(z, n, data) \
+  BOOST_PP_IF(COND(n), EXPAND_GRAD_CASE(n), )
+#define REP_EXPAND_GRAD_TEMPLATE(n) BOOST_PP_REPEAT(n, EXPAND_TEMPLATE_GRAD, ~)
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
```
Comment (reyoung) [inline paddle/operators/expand_op.h]: Do not `using namespace` or type alias in the header file.

File: paddle/operators/expand_op.h [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+      default:
+        PADDLE_ENFORCE(false, "Only support tensor whose rank in [1, 6].");
+    };
+  }
+
+ protected:
+  template <int Rank>
+  void Expand(const framework::ExecutionContext& context) const {
+    auto* in0 = context.Input<Tensor>("X");
+    auto expand_times = context.Attr<std::vector<int>>("expandTimes");
```
Comment (reyoung) [inline paddle/operators/expand_op.h]: `auto -> auto&`

File: paddle/operators/expand_op.h [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    auto* out0 = context.Output<Tensor>("Out");
+    Eigen::DSizes<int, Rank> bcast_dims;
```
Comment (reyoung) [inline paddle/operators/expand_op.h]: Maybe the problem is how to cast vector<int> to Eigen::DSizes<int, Rank>

**Comment by reyoung** [comment]
Related issue #4091

File: paddle/operators/expand_op.h [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+          reshape_dims_vec.push_back(expand_times[i]);
+        } else {
+          reduce_dims_vec.push_back(reshape_dims_vec.size());
+          reshape_dims_vec.push_back(expand_times[i]);
+          reshape_dims_vec.push_back(x_dims[i]);
+        }
+      }
+    }
+
+    int dims = reshape_dims_vec.size() * 6 + reduce_dims_vec.size() - 7;
```
Comment (guoshengCS) [inline paddle/operators/expand_op.h]: 这里不`-7`的话是否`ExpandBackward`中 `Dims / 6 + 1`和 `Dims % 6 + 1`就可以把`+1`去掉了

File: paddle/operators/expand_op.cc [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  ExpandOpMaker(framework::OpProto* proto, framework::OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "Input tensor.");
+    AddOutput("Out", "Expanded result by tiling input X.");
+    AddAttr<std::vector<int>>("expandTimes",
+                              "Expand times for each dimension.");
+    AddComment(R"DOC(
+Expand operator tiles the input by given times number. You should set times
+number for each dimension by providing attribute 'expandTimes'. Rank of input
+tensor should be in [1, 6]. Please draw an attention that size of
```
Comment (jacquesqiao) [inline paddle/operators/expand_op.cc]: why `Rank of input tensor should be in [1, 6]`

File: paddle/operators/expand_op.cc [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+class ExpandOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext& ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X"), "X must be initialized.");
+    std::vector<int> expand_times = Attr<std::vector<int>>("expandTimes");
+    auto* x = ctx.Input<Tensor>("X");
```
Comment (jacquesqiao) [inline paddle/operators/expand_op.cc]: seems x is not used, can use
```
auto x_dims = ctx.Input<Tensor>("X")->dims();
```
instead

File: python/paddle/v2/framework/tests/test_expand_op.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    def test_check_grad(self):
+        self.check_grad(['X'], 'Out')
+
+
+class TestExpandOpRank2(OpTest):
+    def setUp(self):
+        self.op_type = "expand"
+        self.inputs = {'X': np.random.random((12, 14)).astype("float32")}
+        self.attrs = {'expandTimes': [1, 1]}
+        output = np.tile(self.inputs['X'], (1, 1))
```
Comment (jacquesqiao) [inline python/paddle/v2/framework/tests/test_expand_op.py]: I think use all one as expand time is not good for test ~

File: python/paddle/v2/framework/tests/test_expand_op.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    def test_check_grad(self):
+        self.check_grad(['X'], 'Out')
+
+
+class TestExpandOpRank3(OpTest):
+    def setUp(self):
+        self.op_type = "expand"
+        self.inputs = {'X': np.random.random((2, 4, 5)).astype("float32")}
+        self.attrs = {'expandTimes': [1, 1, 1]}
```
Comment (jacquesqiao) [inline python/paddle/v2/framework/tests/test_expand_op.py]: seem as above

**Comment by reyoung** [comment]
Maybe could use this PR #4205, too?

File: python/paddle/v2/framework/tests/test_expand_op.py (hunk 7 above)
Comment (pkuyym) [reply python/paddle/v2/framework/tests/test_expand_op.py]: Yeah, this is a corner case and I will add more robust test cases. Same as bellow.

File: paddle/operators/expand_op.h (hunk 4 above)
Comment (pkuyym) [reply paddle/operators/expand_op.h]: Not really, please consider 6*1 + 6 = 12,  12/6=2 (expect 1)

**Comment by qingqing01** [comment]
Please update this PR and merge it asap.

File: paddle/operators/expand_op.h [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+template <typename Place, typename T>
+class ExpandGradKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* in0 = context.Input<Tensor>("X");
+    auto& expand_times = context.Attr<std::vector<int>>("expandTimes");
+    auto x_dims = in0->dims();
+    std::vector<int> reshape_dims_vec;
+    std::vector<int> reduce_dims_vec;
```
Comment (qingqing01) [inline paddle/operators/expand_op.h]: Please add the comments about how to compute gradients.

File: paddle/operators/expand_op.h [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    }
+  }
+
+ protected:
+  template <int Dims>
+  void ExpandBackward(const framework::ExecutionContext& context,
+                      const std::vector<int>& reshape_dims_vec,
+                      const std::vector<int>& reduce_dims_vec) const {
+    size_t reshape_size = Dims / 6 + 1;
+    size_t reduce_size = Dims % 6 + 1;
```
Comment (qingqing01) [inline paddle/operators/expand_op.h]: use a constant variable to represent 6, do not use 6 directly.

File: python/paddle/v2/framework/tests/test_expand_op.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        self.outputs = {'Out': output}
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_check_grad(self):
+        self.check_grad(['X'], 'Out')
+
+
+class TestExpandOpRank2_2(OpTest):
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_expand_op.py]: The name _1 _2 is not clear.
